File: app/src/main/java/com/mycompany/oreamnos/ShareBottomSheetFragment.java
package com.mycompany.oreamnos;

import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.content.BroadcastReceiver;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.Editable;
import android.text.TextWatcher;
import android.transition.AutoTransition;
import android.transition.TransitionManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.card.MaterialCardView;
import com.google.android.material.chip.Chip;
import com.google.android.material.progressindicator.LinearProgressIndicator;
import com.google.android.material.textfield.TextInputEditText;
import com.mycompany.oreamnos.services.ContentGenerationService;
import com.mycompany.oreamnos.services.GeminiService;
import com.mycompany.oreamnos.services.WebContentExtractor;
import com.mycompany.oreamnos.utils.HapticHelper;
import com.mycompany.oreamnos.utils.NotificationHelper;
import com.mycompany.oreamnos.utils.PreferencesManager;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Bottom Sheet Dialog Fragment for handling shared content.
 * Features: collapsible input, progress bar, tone toggle, haptic feedback,
 * background processing.
 */
public class ShareBottomSheetFragment extends BottomSheetDialogFragment {

    private static final String ARG_SHARED_TEXT = "shared_text";

    // Views
    private MaterialCardView inputCard;
    private View inputHeader;
    private ImageView expandArrow;
    private TextView sharedText;
    private TextView inputCharCount;
    private View toneToggleContainer;
    private Chip chipFormal;
    private Chip chipCasual;
    private MaterialCardView progressCard;
    private LinearProgressIndicator progressBar;
    private TextView progressStatusText;
    private MaterialCardView resultCard;
    private TextInputEditText outputText;
    private TextView outputWordCount;
    private TextView editedIndicator;
    private MaterialButton editButton;
    private MaterialButton copyButton;
    private MaterialButton shareButton;
    private MaterialButton backgroundButton;
    private MaterialButton continueButton;
    private Chip includeTitleCheckbox;
    private Chip includeHashtagsCheckbox;
    private Chip includeSourceCheckbox;

    // State
    private String originalSharedContent = "";
    private String lastGeneratedPost = "";
    private String generatedSourceCitation = "";
    private String generatedTitle = "";
    private String generatedBody = "";
    private boolean isInputExpanded = true;
    private boolean isEditMode = false;
    private boolean isProcessing = false;

    // Utils
    private PreferencesManager prefsManager;
    private HapticHelper hapticHelper;
    private NotificationHelper notificationHelper;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    /**
     * BroadcastReceiver for background processing results.
     */
    private final BroadcastReceiver serviceResultReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            boolean success = intent.getBooleanExtra(ContentGenerationService.EXTRA_SUCCESS, false);
            if (success) {
                String result = intent.getStringExtra(ContentGenerationService.EXTRA_RESULT);
                handleGenerationSuccess(result);
            } else {
                String error = intent.getStringExtra(ContentGenerationService.EXTRA_ERROR);
                handleGenerationError(error);
            }
        }
    };

    public static ShareBottomSheetFragment newInstance(String sharedText) {
        ShareBottomSheetFragment fragment = new ShareBottomSheetFragment();
        Bundle args = new Bundle();
        args.putString(ARG_SHARED_TEXT, sharedText);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(STYLE_NORMAL, R.style.Theme_Oreamnos_BottomSheet);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_share_bottom_sheet, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Context context = requireContext();
        prefsManager = new PreferencesManager(context);
        hapticHelper = new HapticHelper(context);
        notificationHelper = new NotificationHelper(context);

        initViews(view);
        setupListeners();
        loadSharedContent();
        setupToneToggle();
        setupChipsVisibility();
    }

    @Override
    public void onResume() {
        super.onResume();
        LocalBroadcastManager.getInstance(requireContext()).registerReceiver(
                serviceResultReceiver,
                new IntentFilter(ContentGenerationService.BROADCAST_RESULT));
    }

    @Override
    public void onPause() {
        super.onPause();
        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(serviceResultReceiver);
    }

    @Override
    public void onDismiss(@NonNull android.content.DialogInterface dialog) {
        super.onDismiss(dialog);
        // Notify host activity to finish
        if (getActivity() instanceof ShareReceiverActivity) {
            ((ShareReceiverActivity) getActivity()).onBottomSheetDismissed();
        }
    }

    private void initViews(View view) {
        // Input card
        inputCard = view.findViewById(R.id.inputCard);
        inputHeader = view.findViewById(R.id.inputHeader);
        expandArrow = view.findViewById(R.id.expandArrow);
        sharedText = view.findViewById(R.id.sharedText);
        inputCharCount = view.findViewById(R.id.inputCharCount);

        // Tone toggle
        toneToggleContainer = view.findViewById(R.id.toneToggleContainer);
        chipFormal = view.findViewById(R.id.chipFormal);
        chipCasual = view.findViewById(R.id.chipCasual);

        // Progress
        progressCard = view.findViewById(R.id.progressCard);
        progressBar = view.findViewById(R.id.progressBar);
        progressStatusText = view.findViewById(R.id.progressStatusText);

        // Result
        resultCard = view.findViewById(R.id.resultCard);
        outputText = view.findViewById(R.id.outputText);
        outputWordCount = view.findViewById(R.id.outputWordCount);
        editedIndicator = view.findViewById(R.id.editedIndicator);
        editButton = view.findViewById(R.id.editButton);
        copyButton = view.findViewById(R.id.copyButton);
        shareButton = view.findViewById(R.id.shareButton);
        includeTitleCheckbox = view.findViewById(R.id.includeTitleCheckbox);
        includeHashtagsCheckbox = view.findViewById(R.id.includeHashtagsCheckbox);
        includeSourceCheckbox = view.findViewById(R.id.includeSourceCheckbox);

        // Buttons
        backgroundButton = view.findViewById(R.id.backgroundButton);
        continueButton = view.findViewById(R.id.continueButton);

        // Close button
        ImageButton closeButton = view.findViewById(R.id.closeButton);
        closeButton.setOnClickListener(v -> dismiss());
    }

    private void setupListeners() {
        // Input card collapse/expand
        inputHeader.setOnClickListener(v -> toggleInputCard());
        inputCard.setOnClickListener(v -> {
            if (!isInputExpanded)
                toggleInputCard();
        });

        // Action buttons
        editButton.setOnClickListener(v -> toggleEditMode());
        copyButton.setOnClickListener(v -> onCopyClick());
        shareButton.setOnClickListener(v -> onShareClick());
        backgroundButton.setOnClickListener(v -> onBackgroundClick());
        continueButton.setOnClickListener(v -> onContinueClick());

        // Output chips
        includeTitleCheckbox.setOnCheckedChangeListener((v, checked) -> rebuildOutputText());
        includeHashtagsCheckbox.setOnCheckedChangeListener((v, checked) -> rebuildOutputText());
        includeSourceCheckbox.setOnCheckedChangeListener((v, checked) -> rebuildOutputText());

        // Output text watcher
        outputText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (isEditMode && !s.toString().equals(lastGeneratedPost)) {
                    editedIndicator.setVisibility(View.VISIBLE);
                } else {
                    editedIndicator.setVisibility(View.GONE);
                }
                String text = s.toString().trim();
                int wordCount = text.isEmpty() ? 0 : text.split("\\s+").length;
                outputWordCount.setText(wordCount + " words");
            }
        });
    }

    private void loadSharedContent() {
        if (getArguments() != null) {
            originalSharedContent = getArguments().getString(ARG_SHARED_TEXT, "");
            sharedText.setText(originalSharedContent);
            inputCharCount.setText(originalSharedContent.length() + " chars");

            // Auto-start processing if API key is set
            if (prefsManager.hasApiKey() && !originalSharedContent.isEmpty()) {
                startGeneration();
            } else if (!prefsManager.hasApiKey()) {
                Toast.makeText(getContext(), R.string.api_key_required, Toast.LENGTH_LONG).show();
            }
        }
    }

    private void setupToneToggle() {
        // Set initial tone based on settings
        boolean isFormal = prefsManager.isFormalTone();
        chipFormal.setChecked(isFormal);
        chipCasual.setChecked(!isFormal);
    }

    private void setupChipsVisibility() {
        // Hashtags visibility
        boolean hasHashtags = !prefsManager.getHashtags().isEmpty();
        includeHashtagsCheckbox.setVisibility(hasHashtags ? View.VISIBLE : View.GONE);
        includeHashtagsCheckbox.setChecked(prefsManager.areHashtagsEnabled());

        // Source visibility
        boolean isSourceEnabled = prefsManager.isSourceEnabled();
        includeSourceCheckbox.setVisibility(isSourceEnabled ? View.VISIBLE : View.GONE);
        includeSourceCheckbox.setChecked(isSourceEnabled);
    }

    private void toggleInputCard() {
        isInputExpanded = !isInputExpanded;

        TransitionManager.beginDelayedTransition((ViewGroup) inputCard.getParent(), new AutoTransition());

        if (isInputExpanded) {
            sharedText.setVisibility(View.VISIBLE);
            sharedText.setMaxLines(8);
            expandArrow.setRotation(180);
        } else {
            sharedText.setVisibility(View.GONE);
            expandArrow.setRotation(0);
        }
    }

    private void collapseInputCard() {
        if (isInputExpanded) {
            toggleInputCard();
        }
    }

    private void startGeneration() {
        if (isProcessing)
            return;
        isProcessing = true;

        // Haptic feedback
        hapticHelper.onGenerationStart();

        // Show progress
        progressCard.setVisibility(View.VISIBLE);
        progressBar.setProgress(0);
        progressStatusText.setText(R.string.progress_extracting);
        toneToggleContainer.setVisibility(View.GONE);
        backgroundButton.setVisibility(View.VISIBLE);

        // Get selected tone
        String tone = chipFormal.isChecked() ? PreferencesManager.TONE_FORMAL : PreferencesManager.TONE_CASUAL;
        boolean includeSource = prefsManager.isSourceEnabled();

        executor.execute(() -> {
            try {
                String textToProcess = originalSharedContent;

                // Check if content is a URL
                if (WebContentExtractor.isUrl(originalSharedContent)) {
                    mainHandler.post(() -> animateProgress(0, 40));
                    WebContentExtractor extractor = new WebContentExtractor();
                    textToProcess = extractor.extractContent(originalSharedContent);
                }

                // Update progress
                mainHandler.post(() -> {
                    animateProgress(40, 80);
                    progressStatusText.setText(R.string.progress_generating);
                });

                // Generate post
                String apiKey = prefsManager.getApiKey();
                String endpoint = prefsManager.getApiEndpoint();
                GeminiService gemini = new GeminiService(apiKey, endpoint, tone);
                String result = gemini.curatePost(textToProcess, includeSource);

                mainHandler.post(() -> {
                    animateProgress(80, 100);
                    mainHandler.postDelayed(() -> handleGenerationSuccess(result), 300);
                });

            } catch (Exception e) {
                mainHandler.post(() -> handleGenerationError(e.getMessage()));
            }
        });
    }

    private void animateProgress(int from, int to) {
        ValueAnimator animator = ValueAnimator.ofInt(from, to);
        animator.setDuration(500);
        animator.addUpdateListener(animation -> progressBar.setProgress((int) animation.getAnimatedValue()));
        animator.start();
    }

    private void handleGenerationSuccess(String result) {
        isProcessing = false;

        // Haptic feedback
        hapticHelper.onGenerationComplete();

        // Extract source and title/body
        String contentWithoutSource = extractSourceCitation(result);
        extractTitleAndBody(contentWithoutSource);

        // Rebuild output
        rebuildOutputText();

        // Update UI
        progressCard.setVisibility(View.GONE);
        resultCard.setVisibility(View.VISIBLE);
        backgroundButton.setVisibility(View.GONE);
        continueButton.setVisibility(View.VISIBLE);

        // Collapse input card
        collapseInputCard();
    }

    private void handleGenerationError(String error) {
        isProcessing = false;

        // Haptic feedback
        hapticHelper.onError();

        progressCard.setVisibility(View.GONE);
        toneToggleContainer.setVisibility(View.VISIBLE);

        String errorMsg = error != null ? error : "Unknown error";
        Toast.makeText(getContext(), getString(R.string.processing_error, errorMsg), Toast.LENGTH_LONG).show();
    }

    private String extractSourceCitation(String fullResult) {
        if (fullResult == null) {
            generatedSourceCitation = "";
            return "";
        }

        String regex = "(?im)^[\\s\\p{Z}]*[*_]*(?:Sumber|Source)[*_]*[\\s\\p{Z}]*[:：].*$";
        java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(regex);
        java.util.regex.Matcher matcher = pattern.matcher(fullResult);

        if (matcher.find()) {
            generatedSourceCitation = matcher.group().trim();
            String contentWithoutSource = fullResult.replaceAll(regex, "").trim();
            return contentWithoutSource.replaceAll("\\n+$", "").trim();
        } else {
            generatedSourceCitation = "";
            return fullResult;
        }
    }

    private void extractTitleAndBody(String content) {
        if (content == null || content.isEmpty()) {
            generatedTitle = "";
            generatedBody = "";
            return;
        }

        String[] parts = content.split("\\n\\n", 2);
        if (parts.length >= 2 && parts[0].length() < 150) {
            generatedTitle = parts[0].trim();
            generatedBody = parts[1].trim();
        } else {
            parts = content.split("\\n", 2);
            if (parts.length >= 2 && parts[0].length() < 150) {
                generatedTitle = parts[0].trim();
                generatedBody = parts[1].trim();
            } else {
                generatedTitle = "";
                generatedBody = content.trim();
            }
        }
    }

    private void rebuildOutputText() {
        StringBuilder textBuilder = new StringBuilder();

        if (includeTitleCheckbox.isChecked() && !generatedTitle.isEmpty()) {
            textBuilder.append(generatedTitle).append("\n\n");
        }

        textBuilder.append(generatedBody);

        if (includeHashtagsCheckbox.isChecked() && prefsManager.areHashtagsEnabled()) {
            String hashtags = prefsManager.getFormattedHashtags();
            if (!hashtags.isEmpty()) {
                textBuilder.append("\n\n").append(hashtags);
            }
        }

        if (includeSourceCheckbox.isChecked() && !generatedSourceCitation.isEmpty()) {
            textBuilder.append("\n\n").append(generatedSourceCitation);
        }

        String finalText = textBuilder.toString().trim();
        lastGeneratedPost = finalText;
        outputText.setText(finalText);
    }

    private void toggleEditMode() {
        isEditMode = !isEditMode;

        if (isEditMode) {
            outputText.setFocusable(true);
            outputText.setFocusableInTouchMode(true);
            outputText.requestFocus();
            editButton.setText(R.string.save_edit);
            editButton.setIconResource(android.R.drawable.ic_menu_save);
        } else {
            outputText.setFocusable(false);
            outputText.setFocusableInTouchMode(false);
            editButton.setText(R.string.edit_button);
            editButton.setIconResource(android.R.drawable.ic_menu_edit);
            if (outputText.getText() != null) {
                lastGeneratedPost = outputText.getText().toString();
            }
        }
    }

    private String getFinalText() {
        return outputText.getText() != null ? outputText.getText().toString() : "";
    }

    private void onCopyClick() {
        String textToCopy = getFinalText();
        if (textToCopy.isEmpty())
            return;

        hapticHelper.onCopy();

        ClipboardManager clipboard = (ClipboardManager) requireContext().getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("Oreamnos Post", textToCopy);
        clipboard.setPrimaryClip(clip);

        Toast.makeText(getContext(), R.string.copied_to_clipboard, Toast.LENGTH_SHORT).show();
    }

    private void onShareClick() {
        String textToShare = getFinalText();
        if (textToShare.isEmpty())
            return;

        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_TEXT, textToShare);
        startActivity(Intent.createChooser(shareIntent, getString(R.string.share_button)));
    }

    private void onBackgroundClick() {
        // Start background generation service
        String tone = chipFormal.isChecked() ? PreferencesManager.TONE_FORMAL : PreferencesManager.TONE_CASUAL;
        boolean includeSource = prefsManager.isSourceEnabled();

        Intent serviceIntent = new Intent(requireContext(), ContentGenerationService.class);
        serviceIntent.setAction(ContentGenerationService.ACTION_GENERATE);
        serviceIntent.putExtra(ContentGenerationService.EXTRA_INPUT_TEXT, originalSharedContent);
        serviceIntent.putExtra(ContentGenerationService.EXTRA_INCLUDE_SOURCE, includeSource);
        serviceIntent.putExtra("tone_override", tone);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            requireContext().startForegroundService(serviceIntent);
        } else {
            requireContext().startService(serviceIntent);
        }

        Toast.makeText(getContext(), R.string.notification_generating_message, Toast.LENGTH_SHORT).show();
        dismiss();
    }

    private void onContinueClick() {
        Intent intent = new Intent(requireContext(), MainActivity.class);
        intent.putExtra("shared_text", originalSharedContent);
        intent.putExtra("generated_content", lastGeneratedPost);
        intent.putExtra("generated_title", generatedTitle);
        intent.putExtra("generated_body", generatedBody);
        intent.putExtra("generated_source", generatedSourceCitation);
        startActivity(intent);
        dismiss();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        executor.shutdown();
    }
}

File: app/src/main/java/com/najmi/oreamnos/ShareBottomSheetFragment.kt
package com.najmi.oreamnos

import android.content.BroadcastReceiver
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.Share
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.najmi.oreamnos.services.ContentGenerationService
import com.najmi.oreamnos.services.GeminiService
import com.najmi.oreamnos.services.WebContentExtractor

import com.najmi.oreamnos.ui.theme.ErrorRed
import com.najmi.oreamnos.ui.theme.SocurateTheme
import com.najmi.oreamnos.utils.HapticHelper
import com.najmi.oreamnos.utils.PreferencesManager
import com.najmi.oreamnos.utils.ReadabilityUtils
import com.najmi.oreamnos.utils.StringUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

// File-level regex patterns
private val WHITESPACE_PATTERN = Regex("\\s+")
private val SOURCE_CITATION_PATTERN = Regex("(?im)^[\\s\\p{Z}]*[*_]*(?:Sumber|Source)[*_]*[\\s\\p{Z}]*[:：].*$")
private val TRAILING_NEWLINES_PATTERN = Regex("\\n+$")

// Helper function to extract title and body from content
private fun parseContentTitleAndBody(content: String): Pair<String, String> {
    if (content.isEmpty()) return "" to ""
    val parts = content.split("\n\n", limit = 2)
    if (parts.size >= 2 && parts[0].length < 150) {
        return parts[0].trim() to parts[1].trim()
    }
    val singleParts = content.split("\n", limit = 2)
    if (singleParts.size >= 2 && singleParts[0].length < 150) {
        return singleParts[0].trim() to singleParts[1].trim()
    }
    return "" to content.trim()
}

/**
 * Bottom Sheet Dialog Fragment for handling shared content.
 * Converted to Jetpack Compose.
 */
class ShareBottomSheetFragment : BottomSheetDialogFragment() {

    companion object {
        private const val ARG_SHARED_TEXT = "shared_text"

        fun newInstance(sharedText: String): ShareBottomSheetFragment {
            return ShareBottomSheetFragment().apply {
                arguments = Bundle().apply { putString(ARG_SHARED_TEXT, sharedText) }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NORMAL, R.style.Theme_Oreamnos_BottomSheet)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val prefsManager = PreferencesManager(requireContext())
        val themeMode = prefsManager.getTheme()
        
        return ComposeView(requireContext()).apply {
            setContent {
                SocurateTheme(themeMode = themeMode) {
                    ShareBottomSheetContent(
                        sharedText = arguments?.getString(ARG_SHARED_TEXT) ?: "",
                        onDismiss = { dismiss() },
                        onContinue = { content, title, body, source ->
                            val intent = Intent(requireContext(), MainActivity::class.java).apply {
                                putExtra("shared_text", arguments?.getString(ARG_SHARED_TEXT))
                                putExtra("generated_content", content)
                                putExtra("generated_title", title)
                                putExtra("generated_body", body)
                                putExtra("generated_source", source)
                            }
                            startActivity(intent)
                            dismiss()
                        }
                    )
                }
            }
        }
    }

    override fun onDismiss(dialog: android.content.DialogInterface) {
        super.onDismiss(dialog)
        (activity as? ShareReceiverActivity)?.onBottomSheetDismissed()
    }
}

@Composable
fun ShareBottomSheetContent(
    sharedText: String,
    onDismiss: () -> Unit,
    onContinue: (content: String, title: String, body: String, source: String) -> Unit
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val prefsManager = remember { PreferencesManager(context) }
    val hapticHelper = remember { HapticHelper(context) }

    // State
    var isInputExpanded by remember { mutableStateOf(true) }
    var isLoading by remember { mutableStateOf(false) }
    var loadingText by remember { mutableStateOf("Generating post...") }
    var error by remember { mutableStateOf<String?>(null) }
    var generatedTitle by remember { mutableStateOf("") }
    var generatedBody by remember { mutableStateOf("") }
    var generatedSource by remember { mutableStateOf("") }
    var outputText by remember { mutableStateOf("") }
    var isEditMode by remember { mutableStateOf(false) }
    var isFormalTone by remember { mutableStateOf(prefsManager.isFormalTone()) }
    var includeTitle by remember { mutableStateOf(true) }
    var includeHashtags by remember { mutableStateOf(prefsManager.areHashtagsEnabled()) }
    var includeSource by remember { mutableStateOf(prefsManager.isSourceEnabled()) }

    val hasHashtags = prefsManager.getHashtags()?.isNotEmpty() == true
    val hasResult = generatedBody.isNotEmpty()

    // Rebuild output text when toggles change
    fun rebuildOutput() {
        val builder = StringBuilder()
        if (includeTitle && generatedTitle.isNotEmpty()) {
            builder.append(StringUtils.stripLeadingEmojis(generatedTitle)).append("\n\n")
        }
        builder.append(StringUtils.stripLeadingEmojis(generatedBody))
        if (includeSource && generatedSource.isNotEmpty()) {
            builder.append("\n\n").append(generatedSource)
        }
        if (includeHashtags && prefsManager.areHashtagsEnabled()) {
            val hashtags = prefsManager.getFormattedHashtags()
            if (hashtags.isNotEmpty()) {
                builder.append("\n\n").append(hashtags)
            }
        }
        outputText = builder.toString().trim()
    }

    // Start generation
    fun startGeneration() {
        if (isLoading) return
        isLoading = true
        error = null
        hapticHelper.onGenerationStart()

        scope.launch {
            try {
                var textToProcess = sharedText
                
                if (WebContentExtractor.isUrl(sharedText)) {
                    loadingText = "Extracting content..."
                    textToProcess = withContext(Dispatchers.IO) {
                        WebContentExtractor().extractContent(sharedText)
                    }
                }
                
                loadingText = "Generating post..."
                val tone = if (isFormalTone) PreferencesManager.TONE_FORMAL else PreferencesManager.TONE_CASUAL
                
                val result: String? = withContext(Dispatchers.IO) {
                    val curator = com.najmi.oreamnos.curator.CuratorFactory.create(context)
                    curator.curatePost(textToProcess, prefsManager.isSourceEnabled(), false)
                }
                
                // Extract source citation
                val resultText = result ?: ""
                val matcher = SOURCE_CITATION_PATTERN.find(resultText)
                if (matcher != null) {
                    generatedSource = matcher.value.trim()
                    val contentWithoutSource = SOURCE_CITATION_PATTERN.replace(resultText, "").trim()
                    val parsed = parseContentTitleAndBody(TRAILING_NEWLINES_PATTERN.replace(contentWithoutSource, "").trim())
                    generatedTitle = parsed.first
                    generatedBody = parsed.second
                } else {
                    generatedSource = ""
                    val parsed = parseContentTitleAndBody(resultText)
                    generatedTitle = parsed.first
                    generatedBody = parsed.second
                }
                
                rebuildOutput()
                hapticHelper.onGenerationComplete()
                isInputExpanded = false
                
            } catch (e: Exception) {
                error = e.message
                hapticHelper.onError()
            } finally {
                isLoading = false
            }
        }
    }


    // Auto-start if API key is set
    LaunchedEffect(Unit) {
        if (prefsManager.hasApiKey() && sharedText.isNotEmpty()) {
            startGeneration()
        } else if (!prefsManager.hasApiKey()) {
            Toast.makeText(context, R.string.api_key_required, Toast.LENGTH_LONG).show()
        }
    }

    // Update output when toggles change
    LaunchedEffect(includeTitle, includeHashtags, includeSource) {
        if (hasResult) rebuildOutput()
    }

    Surface(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(topStart = 24.dp, topEnd = 24.dp),
        color = MaterialTheme.colorScheme.surface
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
                .verticalScroll(rememberScrollState())
        ) {
            // Header
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text("Share Content", style = MaterialTheme.typography.titleLarge)
                IconButton(onClick = onDismiss) {
                    Icon(Icons.Default.Close, "Close")
                }
            }
            
            Spacer(Modifier.height(16.dp))
            
            // Input Card (Collapsible)
            Surface(
                modifier = Modifier.fillMaxWidth().clickable { isInputExpanded = !isInputExpanded },
                shape = RoundedCornerShape(12.dp),
                color = MaterialTheme.colorScheme.surfaceVariant
            ) {
                Column(modifier = Modifier.padding(12.dp)) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text("Input", style = MaterialTheme.typography.labelMedium)
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text("${sharedText.length} chars", style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                            Icon(
                                Icons.Default.KeyboardArrowDown,
                                "Expand",
                                modifier = Modifier.rotate(if (isInputExpanded) 180f else 0f)
                            )
                        }
                    }
                    AnimatedVisibility(visible = isInputExpanded, enter = expandVertically(), exit = shrinkVertically()) {
                        Text(
                            text = sharedText,
                            style = MaterialTheme.typography.bodyMedium,
                            maxLines = 8,
                            overflow = TextOverflow.Ellipsis,
                            modifier = Modifier.padding(top = 8.dp)
                        )
                    }
                }
            }
            
            Spacer(Modifier.height(16.dp))
            
            // Tone Toggle (only show before result)
            if (!hasResult && !isLoading) {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    FilterChip(
                        selected = !isFormalTone,
                        onClick = { isFormalTone = false },
                        label = { Text("Casual") }
                    )
                    FilterChip(
                        selected = isFormalTone,
                        onClick = { isFormalTone = true },
                        label = { Text("Formal") }
                    )
                }
                Spacer(Modifier.height(16.dp))
            }
            
            // Loading State
            if (isLoading) {
                Surface(
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    color = MaterialTheme.colorScheme.surfaceVariant
                ) {
                    Column(
                        modifier = Modifier.fillMaxWidth().padding(24.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        CircularProgressIndicator(modifier = Modifier.size(48.dp))
                        Spacer(Modifier.height(16.dp))
                        Text(loadingText, style = MaterialTheme.typography.bodyMedium)
                    }
                }
            }
            
            // Error State
            error?.let { errorMsg ->
                Surface(
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    color = ErrorRed.copy(alpha = 0.1f)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text("Generation Failed", style = MaterialTheme.typography.titleMedium, color = ErrorRed)
                        Spacer(Modifier.height(8.dp))
                        Text(errorMsg, style = MaterialTheme.typography.bodyMedium)
                        Spacer(Modifier.height(12.dp))
                        Button(onClick = { startGeneration() }) { Text("Try Again") }
                    }
                }
            }
            
            // Result
            if (hasResult && error == null && !isLoading) {
                // Output toggles
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    FilterChip(selected = includeTitle, onClick = { includeTitle = !includeTitle }, label = { Text("Title") })
                    if (hasHashtags) {
                        FilterChip(selected = includeHashtags, onClick = { includeHashtags = !includeHashtags }, label = { Text("Hashtags") })
                    }
                    if (prefsManager.isSourceEnabled()) {
                        FilterChip(selected = includeSource, onClick = { includeSource = !includeSource }, label = { Text("Source") })
                    }
                }
                
                Spacer(Modifier.height(12.dp))
                
                // Output text
                OutlinedTextField(
                    value = outputText,
                    onValueChange = { if (isEditMode) outputText = it },
                    modifier = Modifier.fillMaxWidth(),
                    enabled = isEditMode,
                    minLines = 5,
                    maxLines = 12
                )
                
                // Word count and readability
                Row(
                    modifier = Modifier.fillMaxWidth().padding(vertical = 8.dp),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    val wordCount = if (outputText.isEmpty()) 0 else WHITESPACE_PATTERN.split(outputText).size
                    val gradeLevel = ReadabilityUtils.calculateFleschKincaidGradeLevel(outputText)
                    Text("$wordCount words", style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    Text("Grade: ${String.format("%.1f", gradeLevel)}", style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
                
                // Action buttons
                Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    OutlinedButton(onClick = { isEditMode = !isEditMode }, modifier = Modifier.weight(1f)) {
                        Text(if (isEditMode) "Save" else "Edit")
                    }
                    OutlinedButton(
                        onClick = {
                            val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                            clipboard.setPrimaryClip(ClipData.newPlainText("Socurate Post", outputText))
                            Toast.makeText(context, R.string.copied_to_clipboard, Toast.LENGTH_SHORT).show()
                        },
                        modifier = Modifier.weight(1f)
                    ) { Text("Copy") }
                    Button(
                        onClick = {
                            val intent = Intent(Intent.ACTION_SEND).apply {
                                type = "text/plain"
                                putExtra(Intent.EXTRA_TEXT, outputText)
                            }
                            context.startActivity(Intent.createChooser(intent, "Share"))
                        },
                        modifier = Modifier.weight(1f)
                    ) {
                        Icon(Icons.Default.Share, null, modifier = Modifier.size(18.dp))
                        Spacer(Modifier.width(4.dp))
                        Text("Share")
                    }
                }
                
                Spacer(Modifier.height(12.dp))
                
                // Continue button
                Button(
                    onClick = { onContinue(outputText, generatedTitle, generatedBody, generatedSource) },
                    modifier = Modifier.fillMaxWidth()
                ) { Text("Continue in App") }
            }
            
            Spacer(Modifier.height(16.dp))
        }
    }
}
